pkg/kubectl/cmd/completion/completion.go

```diff
@@ -312,3 +317,61 @@ _complete kubectl 2>/dev/null
 	out.Write([]byte(zshTail))
 	return nil
 }
+
+// addPluginCommand adds plugin commands under the given command so that
+// completion includes them
+func addPluginCommands(kubectl *cobra.Command) {
+	output := &bytes.Buffer{}
+	streams := genericclioptions.IOStreams{
+		In:     &bytes.Buffer{},
+		Out:    output,
+		ErrOut: ioutil.Discard,
+	}
+
+	o := &plugin.PluginListOptions{
+		IOStreams: streams,
+		NameOnly:  true,
+	}
+	o.Complete(kubectl)
+	o.Run()
+
+	plugins := strings.Split(output.String(), "\n")
+	for _, plugin := range uniquePluginsList(plugins) {
+		var cmds []*cobra.Command
+
+		// Plugins are named "kubectl-<name>" or with more - such as
+		// "kubectl-<name>-<subcmd1>..."
+		for i, cmdName := range strings.Split(plugin, "-")[1:] {
```

Comment:
I think there is a duplication problem here where plugins with subcommands are causing a problem. It's hard to explain so I'll use an example that broke things for me.  Say we have:
```
kubectl-aa
kubectl-aa-bb
```
The loop at line 344 will first process `kubectl-aa` and will add the command `aa` to the root command (in `_kubectl_root_command()` of the completion output).  This will also create the method `_kubectl_aa()` in the resulting completion output with no subcommands.

The next time the loop at line 344 is called it will be for `kubectl-aa-bb`.  This will add command `aa` a second time to the root command (and again in `_kubectl_root_command()`) and will create method `_kubectl_aa()` a **second** time in the completion output but this time with subcommand `bb`.

This may seem simply inefficient but it some cases it actually breaks things.  For me, when executing this example, the method  `_kubectl_aa()` with `bb` as subcommand is present first, and gets overwritten by `_kubectl_aa()` with no subcommands; so completion does not work for `kubectl aa <TAB>`

In this case the completion output looks like this:
```
[...]

_kubectl_aa()           ### First instance of this method
{
    last_command="kubectl_aa"
    commands=()
    commands+=("bb")

[...]

_kubectl_aa()           ### Second instance of this method overwrites first
{
    last_command="kubectl_aa"
    commands=()
    commands+=()   ### No subcommand in the second instance!

[...]

_kubectl_root_command()
{
    last_command="kubectl"
    commands=()
    commands+=("aa")
    commands+=("aa")        ##### Not broken, but duplicate
    commands+=("annotate")
[...]
```
I think that we need a method, say `longestFormOfPlugin()` which will take `kubectl-aa` and `kubectl-aa-bb` and only keep the longest one.
That will avoid duplication.

Reply:
@marckhouzam I fixed the problem at 6828fe05ff. Would you please check the fix?

Preparing:
```
$ export PATH="~/bin:$PATH"
$ touch ~/bin/kubectl-aa ~/bin/kubectl-aa-bb
$ chmod +x ~/bin/kubectl-aa ~/bin/kubectl-aa-bb
$ _output/bin/kubectl plugin list --name-only 2>/dev/null | grep kubectl-aa
kubectl-aa
kubectl-aa-bb
```

Before (`_kubectl_aa` are duplicated):
```
_output/bin/kubectl completion bash | grep "_kubectl_aa"
_kubectl_aa_bb()
_kubectl_aa()
_kubectl_aa()
```

After:
```
$ _output/bin/kubectl completion bash | grep "_kubectl_aa"
_kubectl_aa_bb()
_kubectl_aa()
``

Reply:
Nicely done!  I tested with different plugins and subplugins and it works great.

One suggestion: I believe that with your use of `kubectl.Find(args)` we no longer need `uniquePluginsList()`. When `kubectl.Find(args)` will hit a duplicate plugin, it will return an empty `remainingArgs` and the duplicate plugin will be ignored.  I think it makes the code simpler to avoid having to define that method, if we don't need it anymore.

Reply:
I modified the code as you mentioned at 960a4cb.